agent/grpc-external/services/peerstream/replication.go

```diff
@@ -351,17 +361,38 @@ func (s *Server) handleUpdateService(
 		}
 		req.Service = nil
 
-		// Then register all checks on that node
+		// Then register all checks on that node - skip the existent ones
 		var chks structs.HealthChecks
 		for _, svcSnap := range nodeSnap.Services {
 			for _, c := range svcSnap.Checks {
-				chks = append(chks, c)
+				exist = false
+				for _, stored := range storedInstances {
+					if stored.Service.Service != svcSnap.Service.Service {
+						continue
+					}
+
+					for _, chk := range stored.Checks {
```

Comment:
Is there another, more performant way we could do this? I worry about 4 nested-loops being a performance issue in some situations.

Reply:
Agreed; one option is to preload a map using the key of combined server and check ID and reuse it for every check.